ngcc/i18n - legacy message id format

petebacondarwin opened the pull request:
This PR is triggered by the use case given in https://github.com/angular/angular/issues/34056

The idea is that ngcc will now always (unless explicitly turned off) add all the legacy message ids to templates that it compiles. Then we can try to use these for matching translations when we get to actually merging the translations at compile-time.

(The first few commits are just general tidy up and not necessary for this PR)

gkalpak reviewed (approved):
LGTM for `fw-ngcc`.
Nice, managable commits :100: :heart:

josephperrott commented on packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts:
```diff
@@ -3457,6 +3457,38 @@ describe('i18n support in the template compiler', () => {
+        var $I18N_0$;
+        if (ngI18nClosureMode) { … }
+        else {
+            $I18N_0$ = $localize \`:␟ec93160d6d6a8822214060dd7938bf821c22b226␟6795333002533525253:Some Message\`;
```
Just to sanity check since we have had weirdness with encodings after clang format ran, `␟` is the expected character in this string?

petebacondarwin replied:
:-) Yes this is the `LEGACY_ID_INDICATOR` - see https://github.com/angular/angular/pull/34135/files#diff-57fe4c8c3c329d46f463b846ab9475faR47

josephperrott reviewed (approved):
LGTM, on little thing I wanted to check to make sure about encoding.

AndrewKushnir reviewed (approved):
LGTM 👍

AndrewKushnir commented:
- [VE Presubmit](http://test/OCL:283368908:BASE:283368954:1575311745111:5779a7a5)
- [Ivy Presubmit](https://test.corp.google.com/OCL:283368908:BASE:283369130:1575311794658:5b8738c1)

petebacondarwin commented:
Merge-assistance: the integration test failure is not related (a problem with terser dependencies).

petebacondarwin committed:
refactor(ngcc): store whether to render legacy i18n message ids in the bundle

Placing this configuration in to the bundle avoids having to pass the
value around through lots of function calls, but also could enable
support for different behaviour per bundle in the future.

petebacondarwin committed:
test: do not copy address to the clipboard in integration test

petebacondarwin committed:
test: do not store generated i18n files in git

petebacondarwin committed:
fix(ngcc): render legacy i18n message ids by default

By ensuring that legacy i18n message ids are rendered into the templates
of components for packages processed by ngcc, we ensure that these packages
can be used in an application that may provide translations in a legacy
format.

Fixes #34056

petebacondarwin committed:
refactor(compiler): i18n - render legacy i18n message ids

Now that `@angular/localize` can interpret multiple legacy message ids in the
metablock of a `$localize` tagged template string, this commit adds those
ids to each i18n message extracted from component templates, but only if
the `enableI18nLegacyMessageIdFormat` is not `false`.

petebacondarwin committed:
refactor(ivy): i18n - store legacy message ids separately to custom ids

This change will enable the Angular compiler to provide these legacy
message ids by default, which will solve problems with ngcc not knowing
whether to generate legacy ids or not.

petebacondarwin committed:
refactor(ivy): i18n - remove hack around TS tagged literals

When first written there was no way to specify the raw text when
programmatically creating a template tagged literal AST node.

This is now fixed in TS and so the hack is no longer needed.

petebacondarwin committed:
test(ngcc): tidy up helper function

Thanks to @gkalpakl for the better regular expression approach.
